builder/beacon_client.go

```diff
@@ -51,14 +51,36 @@ func (b *BeaconClient) isValidator(pubkey PubkeyHex) bool {
 }
 
 func (b *BeaconClient) getProposerForNextSlot(requestedSlot uint64) (PubkeyHex, error) {
-	/* Only returns proposer if requestedSlot is currentSlot + 1, would be a race otherwise */
+	b.updateValidatorsMap(requestedSlot)
+
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	if b.currentSlot+1 != requestedSlot {
-		return PubkeyHex(""), errors.New("slot out of sync")
+	nextSlotProposer, found := b.slotProposerMap[requestedSlot]
+	if !found {
+		log.Error("inconsistent proposer mapping", "currentSlot", b.currentSlot, "slotProposerMap", b.slotProposerMap)
+		return PubkeyHex(""), errors.New("inconsistent proposer mapping")
+	}
+	return nextSlotProposer, nil
+}
+
+func (b *BeaconClient) updateValidatorsMap(nextSlot uint64) error {
+	b.mu.Lock()
```

Comment:
I think it'd be good to avoid locking `b.mu` twice, and instead require that `updateValidatorsMap` is called with it locked.